### DISCUSSION TITLE
Duo 2 stays active in sleep mode, is it normal?

### DISCUSSION THREAD

**Discussion by JayBee-git** [Archfx/duo-de#55] [General]
Before submitting that as an issue, I wanted to make sure it's not a known problem.
I just installed everything on my Duo 2, so everything is fresh, no extra app installed, just logged my google account and changed some tweaks (removed the center gap for instance). 
The issue I have is simple; while in sleep mode (Duo 2 closed), it becomes noticeably warm and uses a lot of battery, like it's still fully active. It does this even in plane mode without any background app, even after a fresh power cycle.
Is that an expected behavior or a bug?

**Comment by Archfx** [comment]
It is **not normal** actually, what are the settings for display turn-off duration? Other than the center gap, what are the other features that you have enabled?

**Comment by JayBee-git** [comment]
Thanks for the answer!
The screen timeout is left as default, 30 seconds.
Aside from the hinge gap, I enabled the dynamic fps option (I need that for some app/games that don't work well at 90fps), allow setting brightness to the lowest brightness, and I changed the preferred network protocols to match my 5G network (but removed my sim and left it to plane mode since).
Maybe the dynamic fps could cause that? I'll try disabling it just in case.

**Comment by JayBee-git** [comment]
I disabled everything, did a reboot and a full charge, and it's still staying warm on sleep and emptying the battery fast. I'll try doing a factory reset to see what is happening.

**Comment by JayBee-git** [comment]
Out of curiosity I checked the temp with a thermal camera, it seems to be warming up and staying at around 35°C.
Definitely not normal for a device in sleep mode.
![img_thermal_1726142365645](https://github.com/user-attachments/assets/704e073d-e5b3-40c4-b2e4-9924a467db87)
![img_thermal_1726142385085](https://github.com/user-attachments/assets/fcff0b66-a053-4cc6-a484-1c4c9fd67f8d)

**Reply by Archfx** [reply]
Wow, that is not normal. I have a leather skin on my device, still, it settles in this temperature range during use, but not idle. How is your battery life? Did you move from stock 12L? Do you have WOA installed in parallel?

**Comment by JayBee-git** [comment]
My Duo 2 doesn't have any case beside the official bumper. My suspicion is that it never goes into true sleep mode. 
I came from stock 12L, and I have nothing else installed. On the stock 12L the battery was very decent (I got it a few months back new in box), and it didn't warm up during sleep.
Now the battery is almost completely depleting over night (losing about 5~10% per hour in sleep mode).

Yesterday I did a complete factory reset, didn't enter my Google account, updated to latest ROM that just released, put it back to plane mode, changed the hinge gap option only, and charged it to 100%.
This morning it's still warm and battery is down to 20%.
I believe there is really something odd with the sleep mode on the Duo 2 with this ROM.

On the bright side in winter it will be a nice hand warmer 😂

**Reply by Archfx** [reply]
I don't think this is an issue with the ROM but rather a particular issue on your unit. The fact that this is the first time report on such an issue and if this was an issue with the ROM it should have been identified way earlier since this ROM is used by both duo1 and duo2 users.  Also, 20% from 100% is not acceptable for normal operation within a day. Your particular device must have some sort of hardware issue.

**Comment by JayBee-git** [comment]
I understand your point, but then what kind of issue would that be?

The unit is nearly new and I didn't have any single issue of that sort in both android 11 and 12L. Note that It's bought in Japan (I live there), but I don't think this model ever got a japanese specific hardware/ROM.
The battery indicator always had the very common duo 2 issue of staying at low battery for a while before shutting down, giving it a few extra hours, but I've never had the device staying warm in sleep mode, and in plane mode the battery was lasting a few days before being completely drained.

I fail to see how the device staying active in sleep mode could be a hardware issue. Do you have any idea?
If I check the battery stats, I can see 88% of the battery is used by the "system apps". It's the first time I see that on any Android device. There is clearly something in the rom that keeps running in the background, whatever is causing it.
Is there a way to get a log on what process is running there?

**Reply by Archfx** [reply]
Let's take a log and I will compare it with a device that doesn't have this problem. Lock that device and let it go to sleep mode while connected via the USB. Collect the log,


```shell
adb -d logcat  > log.txt
```

Attach it as a file here. Let's see!!

**Reply by Archfx** [reply]
This log should ideally contain only messages related to charging at this moment since it is connected via the cable.

**Reply by JayBee-git** [reply]
Thanks for the instructions, I'll do that and report back asap 👍🏻

**Reply by JayBee-git** [reply]
[log.txt](https://github.com/user-attachments/files/17002546/log.txt)
Here is the log, I let it run for a few minutes to be sure you have plenty of data.

**Comment by JayBee-git** [comment]
Quick question, is the cover stylus charging function on of off by default in this ROM?

**Reply by Archfx** [reply]
Actually, this ROM does not modify your HAL features that are controlled by the Linux subsystem.  So all the sensors and stylus-related things are not changed from the stock.

**Reply by JayBee-git** [reply]
I see. I'm asking that because I do believe I've had the same issue in the past when I enabled that option by mistake.
I left it disabled afterwards so it should in theory stay off.
I do not have the stylus charging case so it didn't work when enabled but it did suck the power and kept the device warm when leaving the device in my bag.

**Comment by JayBee-git** [comment]
I did more testing beside the log file I sent you.

Instead of leaving the Duo 2 in my bag like I always do, I left it on my desk, waiting for the battery to completely die. This time the device isn't staying much warm, but the battery still doesn't last more than a day and half in plane mode and almost 0 on screen time.

Battery stats don't make much sense either for me:
![Screenshot_20240916-232051](https://github.com/user-attachments/assets/594317f2-340f-4e31-8a72-f2696eca8270)
![Screenshot_20240916-232103](https://github.com/user-attachments/assets/546e67f7-697a-4be9-b02b-f1e3dcec241b)

The surface duo 2 has this odd inductive charging coil on the bottom center of the back of the left screen, that when used in combination with the official pen cover, allows to charge the stylus by conducting the power from the main coil to another side coil localted inside the cover itself.
Whenever the duo 2 stylus charge option is on, it's powering that coil even without the pen cover, so my wild guess is that if conductive objects are nearby that might somehow suck more power and make the device warmer. With the og firmware the one time I put it on by mistake, it was also sucking down the battery and keeping the device warm. Which would explain why it's getting warm mostly when in my bag. I have no idea how the charging function could be left in enabled mode, since you said these options should stay on rom change/storage clear.

Of course that could as well be a system app running in the background and depleting the battery.

Is there any way to quickly reinstall the og rom without bricking the device, so that I can do the same test under the same condtions in a stock state? That would make it very easy to see if the problem comes from my hardware or not.

**Reply by Archfx** [reply]
I briefly gone through your device log file, looks like there are lot of background processes even during the sleep mode (assuming that entire log is collected during the device is in sleep mode). It could be a app or setting that you had in the stock ROM that is affecting the A14. 

Example is this, this is some sort of app trying to use your location services when the device in sleep mode.
```shell
09-14 14:47:55.141  2427  2427 E TLOC    : [TLOCChannel]"Can't receive message due to timeout"
09-14 14:47:55.175  2427  2427 I TLOC    : [ModemComm  ]"Request succeeded for gps engine lock"
09-14 14:47:55.175  2427  2427 D TLOC    : [TLOCChannel] "Going to receive message for TLOCD"
09-14 14:47:55.175  2427  2427 D TLOC    : [TLOCChannel] "wait for message for user TLOCD"
09-14 14:47:56.147  2427  2427 E TLOC    : [TLOCChannel]"Can't receive message due to timeout"
09-14 14:47:56.147  2427  2427 E TLOC    : [TLOCDaemon ]"Couldn't receive command response"
09-14 14:47:56.147  2427  2427 I TLOC    : [TLOCDaemon ]"timeout passed - start periodic warm-up"
09-14 14:47:56.183  2427  2427 I TLOC    : [ModemComm  ]"Start a session successfully"
```

Going back to stock ROM [this](https://github.com/Archfx/duo-de/discussions/51) will help you. If you want to test it again, return to stock, factory reset the device and flash this image again. This should solve any background process that is conflicting with the A14 installation.

**Reply by JayBee-git** [reply]
Wait, I think I still don't really understand how this whole thing work, but the GSI I installed, is that a full rom or something installed on top of it. the app using some power, we should be able to access them no? or do I need to go back to the factory rom? note that the stock rom was already reset to factory when I unlocked the bootloader.
Anyweay thanks for the info I'll try to go back to stock and do a big cleanup and disable everything I don't need.

**Reply by Archfx** [reply]
This is what is happening inside,

Treble separates the vendor and system partitions to allow different versions of Android to run without affecting the vendor-specific parts. However, the user data partition is not wiped during the installation of a GSI unless explicitly formatted.

- System partition: This is replaced with the new GSI.
- User partition (/data): Contains app data, user configurations, and settings. If not wiped, the old data remains intact.
- Apps and configurations: Since they reside in the /data partition, they are restored after booting with the new GSI.

To start fresh, you would need to wipe the /data partition during the installation process (e.g., through recovery or fastboot) to remove previous apps and configurations.

**Reply by Archfx** [reply]
Your device may have specific vendor applications such as carrier-related (region-specific) apps and such, they might also cause issues. Looking at logs makes it much harder to figure out which package is triggering which. Yeah better to revert back to stock, and observe the device for a couple of days. You can side-load (alongside the stock) this ROM as another option as well. You can find information about this on the XDA-developers thread.

**Reply by JayBee-git** [reply]
Thanks for the explanation.
I did a complete factory reset before installing this ROM (bootloader unlocking required it), and another one after the installation, wouldn't that clear the /data partition anyway?

**Comment by JayBee-git** [comment]
I flashed the stock rom back, did a full factory reset, loaded the duo 2 in the exact same state as the Duo-DE ROM, disabled the stylus charging function (that is enabled by default from factory), charged it to 100%, switched to plane mode, unplugged it, and left it at the usual place to let it sleep.
And as expected, in 24 hours the battery only lost 6%, compared to the 60% it loses with Duo-DE in the same amount of time.
I am going to continue the same test with the stylus charging function enabled, to see if it changes anything and keep you updated.

**Comment by JayBee-git** [comment]
I left it 24h in my bag with the stylus charging option on, and as expected it did use more battery (about 16% instead of the 6% yesterday). But still very far from what I get with Duo-DE, and it doesn't get warm at all.
It seems to stay a lot cooler with the stock ROM, so indeed there was something in Duo-DE sucking the battery.

Now the question is how can I avoid this issue when re flashing Duo-DE? 
Last time I did a complete factory reset before flash (I had to, since unlocking the bootloader resets everything), and I did a few factory resets while in the ROM, to no avail, so I'm not sure what to try this time.
Can you please tell me what process I should follow?

**Comment by Archfx** [comment]
Okay, we have two options, let's start with the first one, use the below commands and flash the latest image,

```shell
./adb reboot fastboot
./fastboot delete-logical-partition system_ext_a
./fastboot delete-logical-partition system_ext_b
./fastboot delete-logical-partition product_a
./fastboot delete-logical-partition product_b

./fastboot flash system aosp-arm64-ab-gapps14.0-<<latest>>.img
./fastboot reboot 
```

Let's see how it goes, if not let's try the second option.

**Reply by JayBee-git** [reply]
Hmm that looks exactly like what I did the first time, but ok let me try that out 👍

**Comment by JayBee-git** [comment]
I flashed the ROM using your fastboot commands (then I had to do a factory reset since it wouldn't boot without the ext and product partition), and it seems to be sucking the battery at the same pace as last time.

Btw was this ROM ever tested to check the battery depletion in sleep mode? Would it be possible for you to leave your working Duo 2 to sleep in plane mode after a charge and tell me how quickly the battery goes down?
Just to have a good comparison reference.

**Reply by JayBee-git** [reply]
I did more testing with the ROM, with various settings. When using the device a lot the screen on time battery usage seems normal, but the battery usage in sleep mode is still terrible.
With 0 screen on time in plane mode with nothing running in the background, I barely get more than a day and half, compared to a full week with the og ROM.

It would really help if someone else could test the battery usage in sleep mode on their Duo 2, I can't keep doing blind testing if we don't have a base reference point from other devices.

**Comment by JayBee-git** [comment]
Hey there, so what was the second option you were talking about? And how is the battery on your own unit?

**Reply by Archfx** [reply]
Please check on the latest image, I have created a redundant sleep mode command to trigger when you close the Duo, the original sleep mode was triggered by the `hall sensor`, if the sensor is not reading correctly it could end up not getting into sleep mode unless you press the power button implicitly. This should put the device into sleep mode when the phone is folded shut. Let's see.

The second method I mentioned is a sideloading method. Prior to that check this one.

**Reply by JayBee-git** [reply]
Thanks, but I can already tell you it's not coming from that. I check every time that the duo 2 went into sleep mode (easy to check when closed with the side "notification" bar), and I did all my recent tests with the Duo folded close on a table with nothing around to mess with it. Even if I was not folding it, it should stll go into sleep mode after 30 seconds anyway.
And in the battery stats, the screen on time is generally 0~3 minutes in total when the battery is almost empty, just the time I woke up the device to check the battery once in a while.

What would be useful to know on the other hand is the baseline battery life expected from this ROM when in sleep&plane modes.
Without this information we have no way of knowing if the tests we do here are a waste of time or not.

**Reply by Archfx** [reply]
Baseline battery status should be about 3-4 days if you don't use it and keep it in standby mode, this might be lower compared to stock since A14 has [Android Find My Device](https://www.google.com/android/find) feature. If you use it with normal usage, it would reduced to about 1 day. The device getting warmer is not expected at all. Sorry, at this point nothing else I can think of improving, especially since this behavior is only in your device and I am **not affiliated** with Microsoft to know about any of the region-specific details and versions of this device. And also your device may have a faulty sensor or such issue, that is used by A14. So options are really slim here.

Still, if you want to test, you can try flashing [treble_aosp](https://github.com/ponces/treble_aosp) from ponces, which this ROM is based on. It will not have any duo-specific features. If the behavior is the same, we don't have any option.

So instead of performing experiments, please try side-loading if you still want to use this ROM. Check the XDA developers thread of A14 Surface Duo to find the steps.

**Reply by JayBee-git** [reply]
Thanks but I don't believe the issue is coming from my device in any way. All the tests we did until now show it's not it. The ROM as well as the hardware itself is the exact same as international, as the device is lacking any region specific thing. I double checked that last week when restoring the stock ROM. And if it was a hardware thing, it would have issues in the stock ROM as well.

My guess here is just that leaving the device on sleep+plane mode without using it at all, until the battery dries out wasn't really tested on a fresh install like I do, and I'm pretty sure if you do the exact same test now you'll get similar results. Because yes, in normal usage it's also one day of battery. Thing is, I am using it only as my secondary device, and leave it in sleep the rest of the time.

As for the device warming up, as I said before it's mostly happening when it's in my bag. Something is most likely messing up with either the hall sensor or the stylus charging function, or a similar issue. My bag contains a lot of magnetic attachments. 
But here too, it's not any issues with the stock ROM.

I know you are doing an amazing job for this ROM, and I thank you very much for taking the time to check with me about this, it's greatly appreciated.

I'll stop wasting your time for now and go back to stock, hoping I can find any root module/hack to remove the hinge there, since it's pretty much the only thing I am interested in.

**Comment by Archfx** [comment]
There is good news on this problem. The culprit for this particular issue was identified (Issue #66). Looks like the issue is coming from GSI itself. I will update you soon on this.

**Reply by Archfx** [reply]
If you are not using a sim card with the device, try the following command as an temporary solution

```shell
./adb shell pm uninstall -k --user 0 com.android.phone
```

**Reply by JayBee-git** [reply]
Thanks for the update, that's good news 👍
I'll try it later on when the update fixes the issue, for now I'm back to stock.
